dbbackup/tests/test_dbbackup.py

```diff
@@ -0,0 +1,161 @@
+    l1.daemon.is_in_log(r'\*\*BROKEN\*\* plugin-dbbackup.py Existing db-backup-file differs from original database', start=needle)
+    l1.daemon.is_in_log(r'UNUSUAL lightningd(.*): JSON-RPC shutdown', start=needle)
+
+
+@unittest.skipIf(not DEVELOPER, "needs DEVELOPER=1")
+def test_dbbackup_recover(node_factory, executor):
+    """Tests db backup plugin, recover from an unfortunate loss of database."""
+
+    # l3 is our unfortunate, may_reconnect=False prevents reconnect-attempts,
+    # but incoming or manual connections still work
+    db_backup = os.path.join(node_factory.directory, "l3_lightningd.sqlite3-backup")
+    opts = [{'may_reconnect': True},
+            {'may_reconnect': False, 'may_fail': True},
+            {'may_reconnect': False, 'may_fail': True,
+             'plugin': 'tests/dbbackup/dbbackup.py',
+             'db-backup-file': db_backup,
+             'disconnect': ['@WIRE_UPDATE_FULFILL_HTLC']}]
+
+    # l3 looses its database with a beneficial HTLC in flight
+    l1, l2, l3 = node_factory.line_graph(3, opts=opts, wait_for_announce=True)
+    l1.wait_for_route(l3)
+
+    phash = l3.rpc.invoice(123000, 'test_pay', 'description')['payment_hash']
+    route = l1.rpc.getroute(l3.info['id'], 123000, 1)['route']
+    l1.rpc.sendpay(route, phash)
+    l3.daemon.wait_for_log('Peer transient failure in CHANNELD_NORMAL')
+
+    # Crash l3 and replace its database with the backup, restart and reconnect
+    # FIXME Make it dev-crash ?
+    l3.daemon.kill()
+    db_orig = os.path.join(l3.daemon.lightning_dir, 'lightningd.sqlite3')
+    os.rename(db_backup, db_orig)
+    l3.daemon.opts.pop('dev-disconnect')
+    l3.daemon.opts.pop('dev-no-reconnect')
+    assert l1.rpc.listsendpays(payment_hash=phash)['payments'][0]['status'] == 'pending'
+    l3.start()
+    l2.rpc.connect(l3.info['id'], 'localhost', l3.port)['id']
+    wait_for(lambda: l1.rpc.listsendpays(payment_hash=phash)['payments'][0]['status'] == 'complete')
+
+    # a HACK to get around `ValueError: 2 nodes had unexpected reconnections`
+    l3.daemon.logs = [re.sub('Peer has reconnected', 'MODDED_PeerHasReconnected', l) for l in l3.daemon.logs]
+    l2.daemon.logs = [re.sub('Peer has reconnected', 'MODDED_PeerHasReconnected', l) for l in l2.daemon.logs]
+
+    # TODO: Can we come up with something harder?
+    # Is option_data_loss_protect: doing anything?
+
+
+def test_dbbackup_write_fail(node_factory, executor):
+    """When the plugin cannot write to backup file for some reason"""
+
+    backup = os.path.join(node_factory.directory, "lightningd.sqlite3-backup")
+    l1 = node_factory.get_node(allow_broken_log=True,
+                               options={'plugin': 'tests/dbbackup/dbbackup.py',
+                                        'db-backup-file': backup})
+
+    # rename backup file will cause db_write failure and crash lightningd
+    os.rename(backup, backup + '_')
```

Comment:
How does this cause a failure writing to the DB? I thought sqlite3 keeps an fd open to the db file, which is unaffected by moves.

Reply:
I have no idea, but it works . The error returned by sqlite3: `attempt to write a readonly database`

Any other ideas how to emulate a disk-full scenario? (maybe loop device)